source/common/http/headers.h

```diff
@@ -5,15 +5,48 @@
 #include "envoy/http/header_map.h"
 
 #include "common/singleton/const_singleton.h"
+#include "common/singleton/threadsafe_singleton.h"
 
 namespace Envoy {
 namespace Http {
 
+// This class allows early override of the x-envoy prefix from bootstrap config,
+// so that servers can configure their own x-custom-string prefix.
+//
+// Once the HeaderValues const singleton has been created, changing the prefix
+// will have no further effect.
+class PrefixValue {
+public:
+  const char* prefix() {
+    absl::WriterMutexLock lock(&m_);
+    read_ = true;
+    return prefix_;
+  }
+  void setPrefix(const char* prefix) {
+    absl::WriterMutexLock lock(&m_);
+    bool same_string = absl::string_view(prefix_) == absl::string_view(prefix);
+    RELEASE_ASSERT(!read_ || same_string,
+                   "Attempting to change the header prefix after it has been used!");
+    // Don't bother swapping if there is no actual change.
+    if (!same_string) {
```

Comment:
Won't this crash in the line above so this if statement will always be false?

Reply:
no, because in the common case it hasn't been read yet.
that said, this is a hack to work around a test-only asan problem - let me handle it more cleanly for prod code.